### PR TITLE
Automate WPJ API Test Cases

### PR DESCRIPTION
**Pull request by p3dr0rv**
Automate [WPJ API Test cases](https://identitydivision.visualstudio.com/Engineering/_testPlans/execute?planId=1262497&suiteId=1262522)  to do less manual testing on each release

### CODE REVIEW

File: azure-pipelines/ui-automation/wpj-test.yml [hunk 1]
```diff
@@ -0,0 +1,175 @@
+# run UI automation testcases for wpj
```
Comment (shahzaibj) [inline azure-pipelines/ui-automation/wpj-test.yml:1]: So this is a completely separate pipeline specifically for WPJ test cases?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/BrokerTestHelper.java [hunk 2]
```diff
@@ -22,41 +22,51 @@
...
 
 import java.util.Arrays;
 import java.util.List;
 
 public class BrokerTestHelper {
-    public static ITestBroker createBrokerFromFlavor(@Nullable final SupportedBrokers supportedBrokersAnnotation) {
+    public static ITestBroker createBrokerFromFlavor(@Nullable final SupportedBrokers supportedBrokersAnnotation, @NonNull final IAppInstaller appSource) {
+        return internalCreateBrokerFromFlavor(supportedBrokersAnnotation, appSource);
+    }
+    public static ITestBroker createBrokerFromFlavor(@Nullable final SupportedBrokers supportedBrokersAnnotation){
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/BrokerTestHelper.java:44]: new line between each method

```suggestion
    }
    
    public static ITestBroker createBrokerFromFlavor(@Nullable final SupportedBrokers supportedBrokersAnnotation){
```

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    @Nullable
+    protected abstract String brokerHostApkName();
+
+    @NonNull
+    @Override
+    public BrokerHost getBrokerHost() {
+        // only initialize once....so calling getBrokerHost from anywhere returns the same instance
+        if (mBrokerHost == null) {
+            final String brokerHostApkName = brokerHostApkName();
+            mBrokerHost = new BrokerHost(brokerHostApkName);
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java:77]:
```suggestion
            mBrokerHost = new BrokerHost(brokerHostApkName());
```

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    /**
+     * Run LabUserQuery to get a cloud device administrator user
+     *
+     * @return cloud device administrator user
+     */
+    public String getLabUserForCloudDeviceAdmin() {
+        final LabUserQuery query = new LabUserQuery();
+        query.userRole = LabConstants.UserRole.CLOUD_DEVICE_ADMINISTRATOR;
+        return LabUserHelper.loadUserForTest(query);
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java:102]: you added the new lab api utilities but still using the old one. Why?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+
+    /**
+     * Build MSAL acquire token parameters with default configuration for WPJ test cases
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java:107]:
```suggestion
     * Build MSAL acquire token parameters with default configuration for WPJ test cases.
```

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        // wpj leave
+        final String wpjState = mBrokerHost.wpjLeave();
+        Assert.assertEquals("Device leave successful.", wpjState);
+        final String upnNull = mBrokerHost.getAccountUpn();
+        Assert.assertNull("Device is not joined", upnNull);
+        Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+        try {
+            final boolean deviceDeleted = LabDeviceHelper.deleteDevice(accountUpn, deviceId);
+        } catch (final LabApiException e) {
+            Assert.assertTrue(e.getCode() == 400);
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java:161]: Why are we looking for 400 here?

File: azure-pipelines/ui-automation/wpj-test.yml (hunk 1 above)
Comment (p3dr0rv) [reply azure-pipelines/ui-automation/wpj-test.yml:1]: Yes, at the time this pipeline was created the broker-release pipeline had not to be created.
I created a [PBI](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1651056) to integrate the wpj pipeline into the broker-release pipeline once this pipeline gets completed
The broker-release pipeline is [still in progress](https://github.com/AzureAD/android-complete/compare/amgusain/broker-ci) Amit picked up the current work and its adding extra functionality.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/BrokerTestHelper.java (hunk 2 above)
Comment (p3dr0rv) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/BrokerTestHelper.java]: 30bc409603c6343679d468a20f055802036cac8d

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java (hunk 3 above)
Comment (p3dr0rv) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java]: solve here 30bc409603c6343679d468a20f055802036cac8d

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java (hunk 5 above)
Comment (p3dr0rv) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java]: I changed, it, I think this way is more clear    
 * Build MSAL acquire token parameters for WPJ test cases using default msal configuration.
     * @see com.microsoft.identity.client.msal.automationapp.R.raw.msal_config_default

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java (hunk 4 above)
Comment (p3dr0rv) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java]: I was using the new API for the jwtParser only, now that you mention I will change it

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java (hunk 6 above)
Comment (p3dr0rv) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java:161]: we are unregistering the device in the previous steps, here we are checking the device has been deleted 
https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/702024c4838e2c6cee32160eecaa07d6f8c1fc09/msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java#L153-L154

Comment (shahzaibj) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java:161]: So if it was already deleted then Lab API throws 400 response?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java [hunk 7]
```diff
@@ -92,19 +98,29 @@ public ITestBroker getBroker() {
...
-     * @return cloud device administrator user
+     * @return cloud device administrator user.
      */
     public String getLabUserForCloudDeviceAdmin() {
-        final LabUserQuery query = new LabUserQuery();
-        query.userRole = LabConstants.UserRole.CLOUD_DEVICE_ADMINISTRATOR;
-        return LabUserHelper.loadUserForTest(query);
+        final LabApiAuthenticationClient authenticationClient =
+                new LabApiAuthenticationClient(BuildConfig.LAB_CLIENT_SECRET);
+        final LabClient labClient = new LabClient(authenticationClient);
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java:108]: We can probably create these as `protected` at the class level

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        }
+    }
+
+
+
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136625.java:166]: lots of extra lines

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136627.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        Assert.assertNull("Device is not joined", upnNull);
+        Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+        try {
+            final boolean deviceDeleted = LabDeviceHelper.deleteDevice(accountUpn, deviceId);
+        } catch (final LabApiException e) {
+            Assert.assertTrue(e.getCode() == 400);
+        }
+    }
+
+
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136627.java:159]: lots of extra lines

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136662.java [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+        Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+        try {
+            final boolean deviceDeleted = LabDeviceHelper.deleteDevice(accountUpn, deviceId);
+        } catch (final LabApiException e) {
+            Assert.assertTrue(e.getCode() == 400);
+        }
+    }
+
+
+
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/TestCase1136662.java:173]: nit: extra lines

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java (hunk 7 above)
Comment (p3dr0rv) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java]: 0cd9fc24a68b9fb8e83bf85dad86af4e8f5c23b3

File: azure-pipelines/ui-automation/wpj-test.yml (hunk 1 above)
Comment (iamgusain) [reply azure-pipelines/ui-automation/wpj-test.yml:1]: Now that broker-release pipeline is up, should that be used run these tests,

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+                new RequestedClaimAdditionalInformation();
+        requestedClaimAdditionalInformation.setEssential(true);
+        // claim {"access_token":{"deviceid":{"essential":true}}}
+        claimsRequest.requestClaimInAccessToken("deviceid", requestedClaimAdditionalInformation);
+        return MsalAuthTestParams.builder()
+                .activity(mActivity)
+                .loginHint(mLoginHint)
+                .scopes(Arrays.asList(mScopes))
+                .promptParameter(Prompt.SELECT_ACCOUNT)
+                .msalConfigResourceId(R.raw.msal_config_instance_aware_common_skip_broker)
```
Comment (iamgusain) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/wpj/AbstractWpjTest.java:152]: skip_broker? Isnt Broker required for this request?

**Pull request closed without merging**